### DISCUSSION TITLE
leafmap add_labels() not working for GeoPandas data

### DISCUSSION THREAD

**Discussion by shi093** [opengeos/leafmap#188] [Q&A] [answered]
If "data" is a GeoPandas dataframe, the following feature does not work, Even if the "data" has the "id" field, the error message says the "data" needs to have a "id" field.
Map.add_labels(data, "id", font_size="12pt", font_color="blue", font_family="arial", font_weight="bold")

However, if I export the "data" to a "json" file and read it back like the following, add_labels() can work fine.
data.to_file("data.json", driver="GeoJSON")
data2 = "data.json"

**Comment by giswqs** [comment, marked as answer]
It has been fixed. Update the package using `leafmap.update_package()` and restart the kernel to take effect.

**Reply by shi093** [reply]
I ran leafmap.update_package() and restarted the kernel, but it's still not working for GeoDataFrame.

**Comment by giswqs** [comment]
Try the following example. It works on my end. If it does not work for you, that means your leafmap is not up to date. If it works, then it is the problem if own dataset. Please provide a sample dataset for debugging. 

```python
import leafmap
import geopandas as gpd

data = "https://raw.githubusercontent.com/giswqs/leafmap/master/examples/data/countries.geojson"
gdf = gpd.read_file(data)

m = leafmap.Map()
m.add_labels(data, "NAME", font_size="12pt", font_color="blue", font_family="arial", font_weight="bold")
m
```

![image](https://user-images.githubusercontent.com/5016453/150707954-1f15d5ce-16c0-4049-8cdd-6aadf0525be2.png)

**Reply by shi093** [reply]
Yes, when I output my data into the geojson file and read it back like in your example, the label worked (as I mentioned earlier). But it doesn't work if I directly use the data like this:

m.add_gdf(gdf_zipcode_selected, layer_name="Zip Code", style=style, hover_style=hover_style, fill_colors = ['red', 'yellow', 'green', 'orange'], info_mode = "on_click", )
m.add_labels(gdf_zipcode_selected, 'zipcode')

my data looks like this:
        zipcode	geom
0	07050	POINT (-74.23485 40.76817)
1	07764	POINT (-74.01850 40.28832)
2	08804	POINT (-75.09525 40.64554)
3	10459	POINT (-73.89289 40.82585)
4	07740	POINT (-73.99184 40.29586)

**Reply by giswqs** [reply]
You can compare the difference between the two GeoDataFrame and find out why it fails. 

Unless you can provide the complete code and sample dataset that can reproduce the error, there's not much I can do here.

**Reply by giswqs** [reply]
Here is another potential workaround.

```python
gdf = gdf_zipcode_selected
gdf['longitude'] = gdf.centroid.x
gdf['latitude'] = gdf.centroid.y
df = leafmap.gdf_to_df(gdf)
m.add_labels(df, 'zipcode')
```

**Comment by shi093** [comment]
I run the code you provided, and got an error message (attached at the end). But it worked if I only add the first three lines of your above code. It seems that the GeoDataFrame variable needs to be in certain format for the "label" to work.
gdf = gdf_zipcode_selected
gdf['longitude'] = gdf.centroid.x
gdf['latitude'] = gdf.centroid.y
m.add_gdf(gdf_zipcode_selected, layer_name="Zip Code", style=style, hover_style=hover_style, fill_colors = ['red', 'yellow', 'green', 'orange'], info_mode = "on_click", )
m.add_labels(gdf_zipcode_selected, 'zipcode')

The error message:

KeyError                                  Traceback (most recent call last)
<ipython-input-67-7dfe671bd5eb> in <module>
     23 gdf['longitude'] = gdf.centroid.x
     24 gdf['latitude'] = gdf.centroid.y
---> 25 df = leafmap.gdf_to_df(gdf)
     26 m.add_labels(df, 'zipcode')
     27 #m.add_labels(gdf_zipcode_selected, 'zipcode', font_size="12pt", font_color="darkblue", font_weight="bold")

~\Miniconda3\envs\geo\lib\site-packages\leafmap\common.py in gdf_to_df(gdf, drop_geom)
   4270 
   4271     if drop_geom:
-> 4272         df = pd.DataFrame(gdf.drop(columns=["geometry"]))
   4273     else:
   4274         df = pd.DataFrame(gdf)

~\Miniconda3\envs\geo\lib\site-packages\pandas\core\frame.py in drop(self, labels, axis, index, columns, level, inplace, errors)
   4306                 weight  1.0     0.8
   4307         """
-> 4308         return super().drop(
   4309             labels=labels,
   4310             axis=axis,

~\Miniconda3\envs\geo\lib\site-packages\pandas\core\generic.py in drop(self, labels, axis, index, columns, level, inplace, errors)
   4151         for axis, labels in axes.items():
   4152             if labels is not None:
-> 4153                 obj = obj._drop_axis(labels, axis, level=level, errors=errors)
   4154 
   4155         if inplace:

~\Miniconda3\envs\geo\lib\site-packages\pandas\core\generic.py in _drop_axis(self, labels, axis, level, errors)
   4186                 new_axis = axis.drop(labels, level=level, errors=errors)
   4187             else:
-> 4188                 new_axis = axis.drop(labels, errors=errors)
   4189             result = self.reindex(**{axis_name: new_axis})
   4190 

~\Miniconda3\envs\geo\lib\site-packages\pandas\core\indexes\base.py in drop(self, labels, errors)
   5590         if mask.any():
   5591             if errors != "ignore":
-> 5592                 raise KeyError(f"{labels[mask]} not found in axis")
   5593             indexer = indexer[~mask]
   5594         return self.delete(indexer)

KeyError: "['geometry'] not found in axis"

**Comment by giswqs** [comment]
The GeoDataFrame must have a `geometry` column. It seems your GeoDataFrame has a column named `geom`, not `geometry`.

**Comment by shi093** [comment]
Thanks for all your help. Really appreciated. 
Here is what I learned through our discussion. As I read the GeoDataFrame from postgresql, and read_postgis() requires there is a "geom" column, but the "label" function requires the geometry column be named "geometry".

sql_cmd = '''select geoid10 as zipcode, ST_Centroid(geom) as geom, ST_Centroid(geom) as geometry from zipcode where geoid10 in ( '''+selected_zip+''' )'''
gdf_zipcode_selected = gpd.read_postgis(sql_cmd, con)

**Reply by giswqs** [reply]
I have updated the `add_labels()` function. It should now work for data with either `geom` or `geometry` columns. Please update the package and try again.
